src/Workspaces/Core/Desktop/Execution/DesktopReferenceSerializationServiceFactory.cs

```diff
@@ -22,10 +29,62 @@ public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
 
         private sealed class Service : AbstractReferenceSerializationService
         {
+            // cache for encoding. 
+            // typical low number, high volumn data cache.
+            private static readonly ConcurrentDictionary<Encoding, byte[]> s_encodingCache = new ConcurrentDictionary<Encoding, byte[]>(concurrencyLevel: 2, capacity: 5);
+
             public Service(ITemporaryStorageService service) : base(service)
             {
             }
 
+            public override void WriteTo(Encoding encoding, ObjectWriter writer, CancellationToken cancellationToken)
+            {
+                if (encoding == null)
+                {
+                    base.WriteTo(encoding, writer, cancellationToken);
+                    return;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                writer.WriteByte(EncodingSerialization);
+
+                byte[] value;
+                if (!s_encodingCache.TryGetValue(encoding, out value))
```

Comment:
imo, because it's much clearer.  There is a method which expresses exactly the concept you're trying to express.  Namely: give me the value if you have it.  Otherwise, compute this value and put it in the map and give me that.